### PR TITLE
Markov Chain Monte Carlo Gaussian splat optimizer

### LINKED ISSUE #181: Implement Python MCMC optimizer subclass
Implement an MCMC subclass of `GaussianSplatOptimizerBase.py` in `reality-capture`.

- Add `refine()`, `step()`, and `zero_grad()` methods, to be called by `GaussianSplatReconstruction.optimize()`
- These Python methods should call into the corresponding fvdb-core MCMC CUDA kernels for optimization
- Ensure compatibility with existing optimizer interfaces in the codebase

Parent issue: #180

### LINKED ISSUE #182: Testing: MCMC optimizer pytests
Develop comprehensive pytests to validate the new MCMC optimizer implementation.

- Cover all methods in the Python MCMC optimizer subclass
- Design tests that verify kernel calls and optimizer behavior
- Add edge cases and expected error conditions
- Integrate tests into the reality-capture CI suite

Parent issue: #180

### LINKED ISSUE #183: Benchmarking: Unit benchmarks for MCMC optimizer
Implement unit benchmarks in `test/benchmark/test_3dgs.py` for regression testing of the MCMC optimizer.

- Develop nightly tests comparing MCMC with existing optimizer benchmarks
- Track both quality and performance metrics
- Ensure benchmarking code is maintainable and produces actionable results

Parent issue: #180

### PR DESCRIPTION
**Pull request by harrism**
Closes #181
Closes #182
Closes #183 
Part of #180

@fwilliams contributed to this PR.

This pull request introduces support for a new MCMC-based optimizer for Gaussian Splat radiance field reconstruction, refactors the optimizer registration and deserialization system for extensibility, and adds shared test utilities for Gaussian Splat models. The changes enable users to select between classic and MCMC optimizer strategies from the `frgs` CLI, ensure optimizer classes are registered and deserialized robustly, and provide reusable testing infrastructure.

it's not possible to train (for example) the garden scene with the MCMC optimizer with a target maximum number of Gaussians of 3M via the following example command. (To train with the "classic" optimizer just change `reconstruct-mcmc` to `reconstruct`).

```
frgs reconstruct-mcmc ./data/360_v2/garden/ --opt.max_gaussians 3000000
```

**New MCMC Optimizer Support**

* Added `GaussianSplatOptimizerMCMC` and `GaussianSplatOptimizerMCMCConfig` classes, and exposed them in the package API (`__init__.py`) for use in reconstruction workflows. [[1]](diffhunk://#diff-2759820a1a42f599112b12750aab136f7c36cbaf9b52cbc188ed359046b0d878R13-R16) [[2]](diffhunk://#diff-2759820a1a42f599112b12750aab136f7c36cbaf9b52cbc188ed359046b0d878R37-R38) [[3]](diffhunk://#diff-9de504483e04f918fb31b67eddc32cbc90c5e2d9353081245e4fbafd9b752912R34-R37)
* Introduced the `ReconstructMCMC` CLI command (`frgs reconstruct-mcmc`), allowing users to reconstruct scenes using the MCMC optimizer strategy via the `frgs` CLI. [[1]](diffhunk://#diff-699808f45eb9286c8cf11740a9eb4479fdf785a555166a3753d492cf9c00e142L15-R33) [[2]](diffhunk://#diff-445e9a7b0b8951b65c2c79fa692acadeb29bd16408c53099fc912d5e4b74ac7aR355-R367)
* Updated the reconstruction pipeline to select the optimizer type (classic or MCMC) based on the optimizer config provided, and improved docstrings to reflect this. [[1]](diffhunk://#diff-9de504483e04f918fb31b67eddc32cbc90c5e2d9353081245e4fbafd9b752912L404-R411) [[2]](diffhunk://#diff-9de504483e04f918fb31b67eddc32cbc90c5e2d9353081245e4fbafd9b752912L439-R446)
* Python tests for the MCMC optimizer and for training using it.
* Basic mcmc unit benchmarks

**Optimizer Registration and Serialization Refactor**

* Implemented a `splat_optimizer` decorator and optimizer registry to ensure all optimizer classes are registered for (de)serialization, and refactored `from_state_dict` to use the registry for robust deserialization. [[1]](diffhunk://#diff-6834ff0b5a5c8f48f5f900b99ed130ed9923ae6a97d5057ed18988a729ce7f8dL5-R36) [[2]](diffhunk://#diff-6834ff0b5a5c8f48f5f900b99ed130ed9923ae6a97d5057ed18988a729ce7f8dR49-R57) [[3]](diffhunk://#diff-6834ff0b5a5c8f48f5f900b99ed130ed9923ae6a97d5057ed18988a729ce7f8dL35-R78) [[4]](diffhunk://#diff-c0c22eb28bebcf7b745fbbc5d31893e0e4682be84e551e3e16d1051f731056c4L19-R19) [[5]](diffhunk://#diff-c0c22eb28bebcf7b745fbbc5d31893e0e4682be84e551e3e16d1051f731056c4R251-R259) [[6]](diffhunk://#diff-c0c22eb28bebcf7b745fbbc5d31893e0e4682be84e551e3e16d1051f731056c4R509)
* Updated serialization to include the optimizer class name in the state dict, supporting round-trip saving/loading across optimizer types. [[1]](diffhunk://#diff-c0c22eb28bebcf7b745fbbc5d31893e0e4682be84e551e3e16d1051f731056c4R509) [[2]](diffhunk://#diff-9de504483e04f918fb31b67eddc32cbc90c5e2d9353081245e4fbafd9b752912L560-R563)

**Testing Infrastructure Improvements**

* Added a `tests/__init__.py` to make the test directory a package, enabling shared test helpers.
* Introduced reusable Gaussian Splat test utilities in `tests/unit/common.py`, including scene loading, model initialization, and rendering helpers, as well as a mixin for test cases.

### CODE REVIEW

File: fvdb_reality_capture/cli/frgs/_reconstruct.py [hunk 1]
```diff
@@ -351,3 +352,16 @@ def execute(self) -> None:
...
+class ReconstructMCMC(Reconstruct):
+    """
+    Reconstruct a Gaussian Splat Radiance Field using the MCMC optimizer strategy.
+
+    This command is identical to :class:`Reconstruct`, but exposes the
+    :class:`~fvdb_reality_capture.radiance_fields.GaussianSplatOptimizerMCMCConfig`
+    configuration under ``--opt.*``.
+    """
+
+    opt: GaussianSplatOptimizerMCMCConfig = field(default_factory=GaussianSplatOptimizerMCMCConfig)
```
[the remark this answers is not in the document]
Comment (harrism) [reply fvdb_reality_capture/cli/frgs/_reconstruct.py]: We don't have tests of `frgs` command line. I don't think my PR description is misleading.

File: fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+
+    @staticmethod
+    def _build_binomial_coeffs(n_max: int, device: torch.device) -> torch.Tensor:
+        """
+        Build a binomial coefficients lookup table of shape [n_max, n_max] as float32.
+        Matches the reference implementation used in fvdb-core tests.
+        """
+        coeffs = torch.zeros((n_max, n_max), device=device, dtype=torch.float32)
+        for row in range(n_max):
+            coeffs[row, 0] = 1.0
```
[the remark this answers is not in the document]
Comment (harrism) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py]: Indeed, this is a future optimization. I will convert this comment into an issue.

File: fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+                break
+        if means_lr is None:
+            raise RuntimeError("Could not find 'means' param group in optimizer")
+        noise_scale = float(self._config.noise_lr) * means_lr
+        if noise_scale != 0.0:
+            self._model.add_noise_to_means(noise_scale=noise_scale)
+
+        self._optimizer.step()
+        self._step_count += 1
+        # Decay the means learning rate
```
[the remark this answers is not in the document]
Comment (harrism) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py]: This is not necessary. This is the same as our other optimizer, and the docstring for this function says "decay the learning rate of the means".

File: fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+            optimizer (torch.optim.Adam): The optimizer for the model.
+            spatial_scale (float): A spatial scale for the scene used to interpret 3D scale thresholds in the config.
+            refine_count (int): The number of times :func:`refine()` has been called on this optimizer.
+            step_count (int): The number of times :func:`step()` has been called on this optimizer.
+            _private (Any): A private object to prevent direct instantiation. Must be
+                :obj:`GaussianSplatOptimizerMCMC.__PRIVATE__`.
+        """
+        if _private is not self.__PRIVATE__:
+            raise RuntimeError(
+                "GaussianSplatOptimizerMCMC must be created using from_model_and_config() or from_state_dict()"
```
[the remark this answers is not in the document]
Comment (harrism) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py]: Fixed.

File: fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+        return GaussianSplatOptimizerMCMC.from_model_and_scene(
+            model=model,
+            sfm_scene=sfm_scene,
+            config=self,
+        )
+
+
+@splat_optimizer
+class GaussianSplatOptimizerMCMC(BaseGaussianSplatOptimizer):
+    """
```
[the remark this answers is not in the document]
Comment (harrism) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py]: Fixed

File: tests/unit/test_gaussian_splat_reconstruction_optimizer_registry.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
+# Copyright Contributors to the OpenVDB Project
+# SPDX-License-Identifier: Apache-2.0
+#
+
+import unittest
+from typing import Any
+
+import torch
+
+import fvdb_reality_capture as frc
```
[the remark this answers is not in the document]
Comment (harrism) [reply tests/unit/test_gaussian_splat_reconstruction_optimizer_registry.py]: Fixed

File: fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py [hunk 7]
```diff
@@ -6,6 +6,7 @@
 import random
 import time
 from dataclasses import dataclass, field
+from enum import Enum
```
[the remark this answers is not in the document]
Comment (harrism) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py]: Fixed

File: tests/unit/test_gaussian_splat_optimizer_mcmc.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+import torch
+from fvdb import GaussianSplat3d
+
+import fvdb_reality_capture as frc
+from tests.unit.common import GettysburgGaussianSplatTestCase
+
+
+class GaussianSplatOptimizerMCMCTests(GettysburgGaussianSplatTestCase, unittest.TestCase):
+    pass
```
[the remark this answers is not in the document]
Comment (harrism) [reply tests/unit/test_gaussian_splat_optimizer_mcmc.py]: Fixed

**Comment by harrism** [comment]
Replaced by #214
